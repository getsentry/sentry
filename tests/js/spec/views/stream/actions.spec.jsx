import React from 'react';
import {mount, shallow} from 'enzyme';

import StreamActions from 'app/views/stream/actions';
import SelectedGroupStore from 'app/stores/selectedGroupStore';

describe('StreamActions', function() {
  let sandbox;
  let actions;
  let wrapper;

  beforeEach(function() {
    sandbox = sinon.sandbox.create();
  });

  afterEach(function() {
    sandbox.restore();
  });

  describe('Bulk', function() {
    describe('Total results > bulk limit', function() {
      beforeAll(function() {
        SelectedGroupStore.records = {};
        SelectedGroupStore.add([1, 2, 3]);
        wrapper = mount(
          <StreamActions
            allResultsVisible={false}
            query=""
            queryCount={1500}
            orgId="1337"
            projectId="1"
            groupIds={[1, 2, 3]}
            onRealtimeChange={function() {}}
            onSelectStatsPeriod={function() {}}
            realtimeActive={false}
            statsPeriod="24h"
          />,
          TestStubs.routerContext()
        );
      });

      it('after checking "Select all" checkbox, displays bulk select message', async function() {
        wrapper.find('ActionsCheckbox Checkbox').simulate('change');
        expect(wrapper.find('.stream-select-all-notice')).toMatchSnapshot();
      });

      it('can bulk select', function() {
        wrapper.find('.stream-select-all-notice a').simulate('click');

        expect(wrapper.find('.stream-select-all-notice')).toMatchSnapshot();
      });

      it('bulk resolves', async function() {
        let apiMock = MockApiClient.addMockResponse({
          url: '/projects/1337/1/issues/',
          method: 'PUT',
        });
        wrapper
          .find('ResolveActions ActionLink')
          .first()
          .simulate('click');

        expect(wrapper.find('ModalDialog')).toMatchSnapshot();
        wrapper.find('Button[priority="primary"]').simulate('click');
        expect(apiMock).toHaveBeenCalledWith(
          expect.anything(),
          expect.objectContaining({
            data: {status: 'resolved'},
          })
        );

        await tick();
        wrapper.update();
      });
    });

    describe('Total results < bulk limit', function() {
      beforeAll(function() {
        SelectedGroupStore.records = {};
        SelectedGroupStore.add([1, 2, 3]);
        wrapper = mount(
          <StreamActions
            allResultsVisible={false}
            query=""
            queryCount={600}
            orgId="1337"
            projectId="1"
            groupIds={[1, 2, 3]}
            onRealtimeChange={function() {}}
            onSelectStatsPeriod={function() {}}
            realtimeActive={false}
            statsPeriod="24h"
          />,
          TestStubs.routerContext()
        );
      });

      it('after checking "Select all" checkbox, displays bulk select message', async function() {
        wrapper.find('ActionsCheckbox Checkbox').simulate('change');
        expect(wrapper.find('.stream-select-all-notice')).toMatchSnapshot();
      });

      it('can bulk select', function() {
        wrapper.find('.stream-select-all-notice a').simulate('click');

        expect(wrapper.find('.stream-select-all-notice')).toMatchSnapshot();
      });

      it('bulk resolves', async function() {
        let apiMock = MockApiClient.addMockResponse({
          url: '/projects/1337/1/issues/',
          method: 'PUT',
        });
        wrapper
          .find('ResolveActions ActionLink')
          .first()
          .simulate('click');

        expect(wrapper.find('ModalDialog')).toMatchSnapshot();
        wrapper.find('Button[priority="primary"]').simulate('click');
        expect(apiMock).toHaveBeenCalledWith(
          expect.anything(),
          expect.objectContaining({
            data: {status: 'resolved'},
          })
        );

        await tick();
        wrapper.update();
      });
    });
  });

  describe('actionSelectedGroups()', function() {
    beforeEach(function() {
      actions = shallow(
        <StreamActions
          query=""
          orgId="1337"
          projectId="1"
          groupIds={[1, 2, 3]}
          onRealtimeChange={function() {}}
          onSelectStatsPeriod={function() {}}
          realtimeActive={false}
          statsPeriod="24h"
        />
      ).instance();
    });

    describe('for all items', function() {
      it("should invoke the callback with 'undefined' and deselect all", function() {
        sandbox.stub(SelectedGroupStore, 'deselectAll');
        let callback = sandbox.stub();

        actions.state.allInQuerySelected = true;

        actions.actionSelectedGroups(callback);

        expect(callback.withArgs(undefined).calledOnce).toBeTruthy();
        expect(SelectedGroupStore.deselectAll.calledOnce).toBeTruthy();

        // all selected is reset
        expect(actions.state.allInQuerySelected).toBe(false);
      });
    });

    describe('for page-selected items', function() {
      it('should invoke the callback with an array of selected items and deselect all', function() {
        sandbox.stub(SelectedGroupStore, 'deselectAll');
        sandbox.stub(SelectedGroupStore, 'getSelectedIds').returns(new Set([1, 2, 3]));

        actions.state.allInQuerySelected = false;
        let callback = sandbox.stub();
        actions.actionSelectedGroups(callback);

        expect(callback.withArgs([1, 2, 3]).calledOnce).toBeTruthy();
        expect(SelectedGroupStore.deselectAll.calledOnce).toBeTruthy();
      });
    });
  });

  describe('missing projectId prop', function() {
    beforeEach(function() {
      wrapper = mount(
        <StreamActions
          query=""
          orgId="1337"
          groupIds={[1, 2, 3]}
          onRealtimeChange={function() {}}
          onSelectStatsPeriod={function() {}}
          realtimeActive={false}
          statsPeriod="24h"
        />,
        TestStubs.routerContext()
      );
    });

    it('should disable resolve picker', function() {
      let resolve = wrapper.find('ResolveActions').first();
      expect(resolve.props().disabled).toBeTruthy();
    });

    it('should disable merge button', function() {
      let merge = wrapper.find('ActionLink[className~="action-merge"]').first();
      expect(merge.props().disabled).toBeTruthy();
    });
  });
});
