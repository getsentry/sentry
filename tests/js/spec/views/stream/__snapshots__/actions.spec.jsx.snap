// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamActions Bulk Total results < bulk limit after checking "Select all" checkbox, displays bulk select message 1`] = `
<div
  className="row stream-select-all-notice"
>
  <div
    className="col-md-12"
  >
    <span>
      3 issues on this page selected.
      <a
        onClick={[Function]}
      >
        <span
          key="5"
        >
          <span
            key="0"
          >
            Select all 
          </span>
          <span
            key="2"
          >
            600
          </span>
          <span
            key="3"
          >
             issues that match this search query.
          </span>
        </span>
      </a>
    </span>
  </div>
</div>
`;

exports[`StreamActions Bulk Total results < bulk limit bulk resolves 1`] = `
<ModalDialog
  bsClass="modal"
  className="in"
  onClick={[Function]}
  role="document"
  style={Object {}}
  tabIndex="-1"
>
  <div
    className="in modal"
    onClick={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
    >
      <div
        className="modal-content"
        role="document"
      >
        <div
          className="modal-body"
        >
          <div>
            <p
              style={
                Object {
                  "marginBottom": "20px",
                }
              }
            >
              <strong>
                <span
                  key="8"
                >
                  <span
                    key="0"
                  >
                    Are you sure you want to 
                  </span>
                  <span
                    key="2"
                  >
                    resolve
                  </span>
                  <span
                    key="3"
                  >
                     all 
                  </span>
                  <span
                    key="5"
                  >
                    600
                  </span>
                  <span
                    key="6"
                  >
                     issues that match the search?
                  </span>
                </span>
              </strong>
            </p>
            <ExtraDescription
              all={true}
              query=""
              queryCount={600}
            >
              <p
                className="error"
              >
                <strong>
                  <span
                    key="5"
                  >
                    <span
                      key="0"
                    >
                      This will apply to all 
                    </span>
                    <span
                      key="2"
                    >
                      600
                    </span>
                    <span
                      key="3"
                    >
                       issues matched in this project!
                    </span>
                  </span>
                </strong>
              </p>
            </ExtraDescription>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <Button
            disabled={false}
            onClick={[Function]}
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <StyledButton
              aria-label="Cancel"
              disabled={false}
              onClick={[Function]}
              role="button"
              style={
                Object {
                  "marginRight": 10,
                }
              }
            >
              <Component
                aria-label="Cancel"
                className="css-cmo08-StyledButton-getColors e17811v30"
                disabled={false}
                onClick={[Function]}
                role="button"
                style={
                  Object {
                    "marginRight": 10,
                  }
                }
              >
                <button
                  aria-label="Cancel"
                  className="css-cmo08-StyledButton-getColors e17811v30"
                  disabled={false}
                  onClick={[Function]}
                  role="button"
                  style={
                    Object {
                      "marginRight": 10,
                    }
                  }
                >
                  <ButtonLabel>
                    <Component
                      className="css-1emshjx-ButtonLabel e17811v31"
                    >
                      <Flex
                        align="center"
                        className="css-1emshjx-ButtonLabel e17811v31"
                      >
                        <Base
                          align="center"
                          className="e17811v31 css-1nhrbny-ButtonLabel"
                        >
                          <div
                            className="e17811v31 css-1nhrbny-ButtonLabel"
                            is={null}
                          >
                            Cancel
                          </div>
                        </Base>
                      </Flex>
                    </Component>
                  </ButtonLabel>
                </button>
              </Component>
            </StyledButton>
          </Button>
          <Button
            autoFocus={true}
            data-test-id="confirm-modal"
            disabled={false}
            onClick={[Function]}
            priority="primary"
          >
            <StyledButton
              aria-label="Bulk resolve issues"
              autoFocus={true}
              data-test-id="confirm-modal"
              disabled={false}
              onClick={[Function]}
              priority="primary"
              role="button"
            >
              <Component
                aria-label="Bulk resolve issues"
                autoFocus={true}
                className="css-1msljz8-StyledButton-getColors e17811v30"
                data-test-id="confirm-modal"
                disabled={false}
                onClick={[Function]}
                priority="primary"
                role="button"
              >
                <button
                  aria-label="Bulk resolve issues"
                  autoFocus={true}
                  className="css-1msljz8-StyledButton-getColors e17811v30"
                  data-test-id="confirm-modal"
                  disabled={false}
                  onClick={[Function]}
                  priority="primary"
                  role="button"
                >
                  <ButtonLabel
                    priority="primary"
                  >
                    <Component
                      className="css-1emshjx-ButtonLabel e17811v31"
                      priority="primary"
                    >
                      <Flex
                        align="center"
                        className="css-1emshjx-ButtonLabel e17811v31"
                      >
                        <Base
                          align="center"
                          className="e17811v31 css-1nhrbny-ButtonLabel"
                        >
                          <div
                            className="e17811v31 css-1nhrbny-ButtonLabel"
                            is={null}
                          >
                            Bulk resolve issues
                          </div>
                        </Base>
                      </Flex>
                    </Component>
                  </ButtonLabel>
                </button>
              </Component>
            </StyledButton>
          </Button>
        </div>
      </div>
    </div>
  </div>
</ModalDialog>
`;

exports[`StreamActions Bulk Total results < bulk limit can bulk select 1`] = `
<div
  className="row stream-select-all-notice"
>
  <div
    className="col-md-12"
  >
    <strong>
      <span
        key="5"
      >
        <span
          key="0"
        >
          Selected all 
        </span>
        <span
          key="2"
        >
          600
        </span>
        <span
          key="3"
        >
           issues that match this search query.
        </span>
      </span>
    </strong>
  </div>
</div>
`;

exports[`StreamActions Bulk Total results > bulk limit after checking "Select all" checkbox, displays bulk select message 1`] = `
<div
  className="row stream-select-all-notice"
>
  <div
    className="col-md-12"
  >
    <span>
      3 issues on this page selected.
      <a
        onClick={[Function]}
      >
        <span
          key="5"
        >
          <span
            key="0"
          >
            Select the first 
          </span>
          <span
            key="2"
          >
            1,000
          </span>
          <span
            key="3"
          >
             issues that match this search query.
          </span>
        </span>
      </a>
    </span>
  </div>
</div>
`;

exports[`StreamActions Bulk Total results > bulk limit bulk resolves 1`] = `
<ModalDialog
  bsClass="modal"
  className="in"
  onClick={[Function]}
  role="document"
  style={Object {}}
  tabIndex="-1"
>
  <div
    className="in modal"
    onClick={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
    >
      <div
        className="modal-content"
        role="document"
      >
        <div
          className="modal-body"
        >
          <div>
            <p
              style={
                Object {
                  "marginBottom": "20px",
                }
              }
            >
              <strong>
                <span
                  key="8"
                >
                  <span
                    key="0"
                  >
                    Are you sure you want to 
                  </span>
                  <span
                    key="2"
                  >
                    resolve
                  </span>
                  <span
                    key="3"
                  >
                     the first 
                  </span>
                  <span
                    key="5"
                  >
                    1,000
                  </span>
                  <span
                    key="6"
                  >
                     issues that match the search?
                  </span>
                </span>
              </strong>
            </p>
            <ExtraDescription
              all={true}
              query=""
              queryCount={1500}
            >
              <p
                className="error"
              >
                <strong>
                  <span
                    key="5"
                  >
                    <span
                      key="0"
                    >
                      This will apply to the first 
                    </span>
                    <span
                      key="2"
                    >
                      1,000
                    </span>
                    <span
                      key="3"
                    >
                       issues matched in this project!
                    </span>
                  </span>
                </strong>
              </p>
            </ExtraDescription>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <Button
            disabled={false}
            onClick={[Function]}
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <StyledButton
              aria-label="Cancel"
              disabled={false}
              onClick={[Function]}
              role="button"
              style={
                Object {
                  "marginRight": 10,
                }
              }
            >
              <Component
                aria-label="Cancel"
                className="css-cmo08-StyledButton-getColors e17811v30"
                disabled={false}
                onClick={[Function]}
                role="button"
                style={
                  Object {
                    "marginRight": 10,
                  }
                }
              >
                <button
                  aria-label="Cancel"
                  className="css-cmo08-StyledButton-getColors e17811v30"
                  disabled={false}
                  onClick={[Function]}
                  role="button"
                  style={
                    Object {
                      "marginRight": 10,
                    }
                  }
                >
                  <ButtonLabel>
                    <Component
                      className="css-1emshjx-ButtonLabel e17811v31"
                    >
                      <Flex
                        align="center"
                        className="css-1emshjx-ButtonLabel e17811v31"
                      >
                        <Base
                          align="center"
                          className="e17811v31 css-1nhrbny-ButtonLabel"
                        >
                          <div
                            className="e17811v31 css-1nhrbny-ButtonLabel"
                            is={null}
                          >
                            Cancel
                          </div>
                        </Base>
                      </Flex>
                    </Component>
                  </ButtonLabel>
                </button>
              </Component>
            </StyledButton>
          </Button>
          <Button
            autoFocus={true}
            data-test-id="confirm-modal"
            disabled={false}
            onClick={[Function]}
            priority="primary"
          >
            <StyledButton
              aria-label="Bulk resolve issues"
              autoFocus={true}
              data-test-id="confirm-modal"
              disabled={false}
              onClick={[Function]}
              priority="primary"
              role="button"
            >
              <Component
                aria-label="Bulk resolve issues"
                autoFocus={true}
                className="css-1msljz8-StyledButton-getColors e17811v30"
                data-test-id="confirm-modal"
                disabled={false}
                onClick={[Function]}
                priority="primary"
                role="button"
              >
                <button
                  aria-label="Bulk resolve issues"
                  autoFocus={true}
                  className="css-1msljz8-StyledButton-getColors e17811v30"
                  data-test-id="confirm-modal"
                  disabled={false}
                  onClick={[Function]}
                  priority="primary"
                  role="button"
                >
                  <ButtonLabel
                    priority="primary"
                  >
                    <Component
                      className="css-1emshjx-ButtonLabel e17811v31"
                      priority="primary"
                    >
                      <Flex
                        align="center"
                        className="css-1emshjx-ButtonLabel e17811v31"
                      >
                        <Base
                          align="center"
                          className="e17811v31 css-1nhrbny-ButtonLabel"
                        >
                          <div
                            className="e17811v31 css-1nhrbny-ButtonLabel"
                            is={null}
                          >
                            Bulk resolve issues
                          </div>
                        </Base>
                      </Flex>
                    </Component>
                  </ButtonLabel>
                </button>
              </Component>
            </StyledButton>
          </Button>
        </div>
      </div>
    </div>
  </div>
</ModalDialog>
`;

exports[`StreamActions Bulk Total results > bulk limit can bulk select 1`] = `
<div
  className="row stream-select-all-notice"
>
  <div
    className="col-md-12"
  >
    <strong>
      <span
        key="5"
      >
        <span
          key="0"
        >
          Selected up to the first 
        </span>
        <span
          key="2"
        >
          1,000
        </span>
        <span
          key="3"
        >
           issues that match this search query.
        </span>
      </span>
    </strong>
  </div>
</div>
`;
